### PR TITLE
[opentitantool] Fix UART timeout issues

### PR DESCRIPTION
**Pull request by cfrantz**
1. Fix timeout detection for the verilator UART.
2. Add timeout detection for the common-case UART.
3. Print backtraces if available.

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

**Review by timothytrippel — APPROVED** [review]
Thanks for fixing this @cfrantz ! I think the `mask_rom_test` example test should now work on Verilator and CW310 with this fix.

File: sw/host/opentitanlib/src/transport/common/uart.rs [hunk 1]
```diff
@@ -64,7 +65,17 @@ impl Uart for SerialPortUart {
     fn read_timeout(&self, buf: &mut [u8], timeout: Duration) -> Result<usize> {
         let mut port = self.port.borrow_mut();
         port.set_timeout(timeout).wrap(UartError::ReadError)?;
-        let len = port.read(buf);
+        let result = port.read(buf);
+        let len = match result {
+            Err(ioerr) => {
```
Comment (jesultra) [inline sw/host/opentitanlib/src/transport/common/uart.rs:70]: How about:
match result {
    Err(ioerr) if ioerr.kind() == ErrorKind::TimedOut => Ok(0),
    _ => result, // Either Ok() or "real" Err()
}

**Comment by jesultra** [comment]
Thanks for this fix.  I wonder how this could have ever worked.  I suppose I failed to test my previous changes, adding the convention that timeout had to be indicated by Ok(0), with HyperDebug, probably testing only with UltraDebug.
Please see my one comment above, which makes the code more concise in my opinion.

File: sw/host/opentitanlib/src/transport/common/uart.rs (hunk 1 above)
Comment (cfrantz) [reply sw/host/opentitanlib/src/transport/common/uart.rs]: Done.

**Pull request merged**